File: clip-it.scala
import scala.util.parsing.combinator._
import java.util.Date
import java.text.SimpleDateFormat

case class Entry(book: Book, metadata: Metadata, highlight: String)
case class Book(title: String, author: String)
case class Metadata(markType: String, where: Location, when: Date)
case class Location(data: String)

class ClipItParser extends JavaTokenParsers {
  override val whiteSpace = """[ \t\r]+""".r
  def eol = "\n"

  def words = rep("""[A-Za-z0-9'":,.+-]+""".r) ^^ (_.mkString(" "))

  def title = words ~ ("(" ~> words <~ ")" <~ eol) ^^ {
    case name ~ author => Book(name, author)
  }

  def meta = "-" ~> ("Bookmark" | "Highlight" | "Note") ~ location ~ ("| Added on" ~> timestamp) ^^ {
    case which ~ loc ~ when => Metadata(which, loc, when)
  }

  def location = words ^^ {
    case data => Location(data)
  }

  def time = """[0-9]+:[0-9]+:[0-9]+""".r

  def timestamp = words ^^ {
    case when =>
      new SimpleDateFormat("EEEE, d MMMM YY HH:mm:ss ZZZZ").parse(when)
  }

  def entry = ((title ~ (meta <~ eol <~ eol) ~ rep(".+".r)) | (title ~ meta)) ^^ {
    case book ~ metadata ~ highlight => Entry(book.asInstanceOf[Book], metadata.asInstanceOf[Metadata], highlight.asInstanceOf[List[String]].mkString(" "))
    case book ~ metadata => Entry(book.asInstanceOf[Book], metadata.asInstanceOf[Metadata], "")
  }
}

object Entrance extends ClipItParser {
  def main: String = {
    val source = scala.io.Source.fromFile("My Clippings.txt")
    val lines = source.mkString.substring(1)
    source.close()
//  val text = "Developing Backbone.js Applications (Addy Osmani)\n - Highlight Loc. 242-43  | Added on Monday, 3 December 12 19:00:12 Greenwich Mean Time\n\n languages when implementing patterns in their projects, there are many lessons we can"

    for (text <- lines.split("==========")) {
      parseAll(entry, text.trim) match {
        case Success(result, _) => println(result)
        case x => {
          println(x)
          return "dicks"
        }
      }

    }

    ""
  }
}

Entrance.main

File: clip-it.py
from datetime import datetime
import re

class Entry:
  def __init__(self, tup):
    self.ent_type = tup[0]

    if ":" in tup[1]:
      colon = tup[1].find(":")
      self.title = tup[1][:colon]
      self.subtitle = tup[1][colon:].strip()
    else:
      self.title = tup[1]
      self.subtitle = ""

    self.author = tup[2]
    self.location = tup[3]
    self.when = tup[4]
    self.highlight = tup[5]


def parse(entry):
  lines = map(lambda x: x.strip(), entry.split("\n"))

  bookmatch = re.search("(.+)\(([^)]*)\)", lines[0])
  if not bookmatch:
    bookmatch = re.search("(.+)-(.+)", lines[0])
  title = bookmatch.group(1).strip()
  author = bookmatch.group(2).strip()

  metamatch = re.search("- ([A-Za-z]+) (.+)\| Added on (.+)", lines[1])
  entrytype = metamatch.group(1)
  location = metamatch.group(2)
  when = datetime.strptime(metamatch.group(3), "%A, %B %d, %Y, %I:%M %p") 

  highlight = ""
  if len(lines) >= 4:
    highlight = "\n".join(lines[3:])

  return Entry((entrytype, title, author, location, when, highlight))


entries = []
with open("My Clippings.txt") as f:
  for entry in f.read()[3:].split("==========")[:-1]:
    entries.append(parse(entry.strip()))

entries = filter(lambda x: x.ent_type == "Highlight", entries)
books = { }


for entry in entries:
  tup = (entry.title, entry.author)
  if tup not in books:
    books[tup] = []
  books[tup].append(entry)


def title_sort(book):
  title = book[0]
  if title[0:4] == "The ":
    return title[4:]
  return title

print "---"
print "layout: page"
print "title: \"Earnest Money\""
print "date: 2013-12-11 00:16"
print "comments: false"
print "sharing: false"
print "footer: true"
print "---"

print ""
for book in sorted(books.keys(), key=title_sort):
  highlights = books[book]
  print "## %s by %s" % book
  print
  for entry in highlights:
    print "* %s" % entry.highlight
  print 
